src/algebra/category/Module/change_of_rings.lean

```diff
@@ -62,4 +75,91 @@ def restrict_scalars {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R 
 @[simp] lemma restrict_scalars.smul_def' {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
   {M : Module.{v} S} (r : R) (m : M) : (r • m : (restrict_scalars f).obj M) = (f r • m : M) := rfl
 
+@[priority 100]
+instance smul_comm_class_mk {R : Type u₁} {S : Type u₂} [ring R] [comm_ring S] (f : R →+* S)
+  (M : Type v) [add_comm_group M] [module S M] :
+  @smul_comm_class R S M ((restrict_scalars.obj' f (Module.mk M)).is_module.to_has_smul) _ :=
+{ smul_comm := λ r s m, (by simp [←mul_smul, mul_comm] : f r • s • m = s • f r • m) }
+
+namespace extend_scalars
+
+open tensor_product
+
+variables {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →+* S)
+
+section unbundled
+
+variables (M : Type v) [add_comm_monoid M] [module R M]
+-- This notation is necessary because we need to reason about `s ⊗ₜ m` where `s : S` and `m : M`;
+-- without this notation, one need to work with `s : (restrict_scalars f).obj ⟨S⟩`.
+localized "notation s `⊗ₜ[` R `,` f `]` m := @tensor_product.tmul R _ _ _ _ _
+  (module.comp_hom _ f) _ s m" in change_of_rings
+
+end unbundled
+
+open_locale change_of_rings
+
+variables (M : Module.{v} R)
+
+/--
+Extension of scalars turn an `R`-module into `S`-module by M ↦ S ⨂ M
+-/
+def obj' : Module S :=
+⟨tensor_product R ((restrict_scalars f).obj ⟨S⟩) M⟩
+
+/--
+Extension of scalars is a functor where an `R`-module `M` is sent to `S ⊗ M` and
+`l : M1 ⟶ M2` is sent to `s ⊗ m ↦ s ⊗ l m`
+-/
+def map' {M1 M2 : Module.{v} R} (l : M1 ⟶ M2) : (obj' f M1) ⟶ (obj' f M2) :=
+@linear_map.base_change R S M1 M2 _ _ ((algebra_map S _).comp f).to_algebra _ _ _ _ l
+
+lemma map'_id {M : Module.{v} R} : map' f (𝟙 M) = 𝟙 _ :=
+linear_map.ext $ λ (x : obj' f M),
+begin
+  dsimp [map'],
+  induction x using tensor_product.induction_on with _ _ m s ihx ihy,
+  { simp only [map_zero], },
+  { rw [linear_map.base_change_tmul, Module.id_apply], },
+  { rw [map_add, ihx, ihy] },
+end
+
+lemma map'_comp {M₁ M₂ M₃ : Module.{v} R} (l₁₂ : M₁ ⟶ M₂) (l₂₃ : M₂ ⟶ M₃) :
+  map' f (l₁₂ ≫ l₂₃) = map' f l₁₂ ≫ map' f l₂₃ :=
+linear_map.ext $ λ (x : obj' f M₁),
+begin
+  dsimp [map'],
+  induction x using tensor_product.induction_on with _ _ x y ihx ihy,
+  { refl, },
+  { tidy, },
+  { simp only [map_add, ihx, ihy], },
+end
+
+end extend_scalars
+
+/--
+Extension of scalars is a functor where an `R`-module `M` is sent to `S ⊗ M` and
+`l : M1 ⟶ M2` is sent to `s ⊗ m ↦ s ⊗ l m`
+-/
+def extend_scalars {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →+* S) :
+  Module R ⥤ Module S :=
```

Comment:
Lean gets the universes right automatically here. Personally, I think that being explicit aids understanding here, but I leave it up to you.
```suggestion
  Module.{v} R ⥤ Module.{max v u₂} S :=
```